primitives/phragmen/fuzzer/src/equalize.rs

```diff
@@ -0,0 +1,157 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Fuzzing fro the equalize algorithm
+//!
+//! It ensures that any solution which gets equalized will lead into a better or equally scored
+//! one.
+
+use honggfuzz::fuzz;
+use sp_phragmen::{
+	equalize, assignment_ratio_to_staked, build_support_map, to_without_backing, elect,
+	PhragmenResult, VoteWeight, evaluate_support, is_score_better,
+};
+use sp_std::collections::btree_map::BTreeMap;
+use sp_runtime::Perbill;
+use rand::{self, Rng, SeedableRng, RngCore};
+
+type AccountId = u64;
+
+fn generate_random_phragmen_result(
+	voter_count: u64,
+	target_count: u64,
+	to_elect: usize,
+	edge_per_voter: u64,
+	mut rng: impl RngCore,
+) -> (PhragmenResult<AccountId, Perbill>, BTreeMap<AccountId, VoteWeight>) {
+	let prefix = 100_000;
+	// Note, it is important that stakes are always bigger than ed and
+	let base_stake: u64 = 1_000_000_000;
+	let ed: u64 = base_stake;
+
+	let mut candidates = Vec::with_capacity(target_count as usize);
+	let mut stake_of_tree: BTreeMap<AccountId, VoteWeight> = BTreeMap::new();
+
+	(1..=target_count).for_each(|acc| {
+		candidates.push(acc);
+		let stake_var = rng.gen_range(ed, 100 * ed);
+		stake_of_tree.insert(acc, base_stake + stake_var);
+	});
+
+	let mut voters = Vec::with_capacity(voter_count as usize);
+	(prefix ..= (prefix + voter_count)).for_each(|acc| {
+		// all possible targets
+		let mut all_targets = candidates.clone();
+		// we remove and pop into `targets` `edge_per_voter` times.
+		let targets = (0..edge_per_voter).map(|_| {
+			let upper = all_targets.len() - 1;
+			let idx = rng.gen_range(0, upper);
+			all_targets.remove(idx)
+		})
+		.collect::<Vec<AccountId>>();
+
+		let stake_var = rng.gen_range(ed, 100 * ed) ;
+		let stake = base_stake + stake_var;
+		stake_of_tree.insert(acc, stake);
+		voters.push((acc, stake, targets));
+	});
+
+	(
+		elect::<AccountId, sp_runtime::Perbill>(
+			to_elect,
+			0,
+			candidates,
+			voters,
+		).unwrap(),
+		stake_of_tree,
+	)
+}
+
+fn main() {
+	let to_range = |x: usize, a: usize, b: usize| {
+		let collapsed = x % b;
+		if collapsed >= a {
+			collapsed
+		} else {
+			collapsed + a
+		}
+	};
```

Comment:
the result can be more than b if `b < a*2`

Reply:
yeah true, this is slightly wrong.

Reply:
for now since all ranges are bigger than the mentioned range, I'll add an assertion to prevent this. I'd be interested to devise a minimum, correct and fair algorithm for this though.